fuse/src/main/java/tachyon/fuse/TachyonFuseFs.java

```diff
@@ -0,0 +1,675 @@
+      return -ErrorCodes.EFAULT();
+    } catch (Throwable e) {
+      LOG.error("Unexpected exception on {}", path, e);
+      return -ErrorCodes.EFAULT();
+    }
+
+    return 0;
+  }
+
+  /**
+   * Flushes cached data on Tachyon.
+   *
+   * Called on explicit sync() operation or at close().
+   * @param path The path on the FS of the file to close
+   * @param fi FileInfo data struct kept by FUSE
+   * @return 0 on success, a negative value on error
+   */
+  @Override
+  public int flush(String path, FuseFileInfo fi) {
+    LOG.trace("flush({})", path);
+    final long fd = fi.fh.get();
+    OpenFileEntry oe = null;
+    synchronized (mOpenFiles) {
+      oe = mOpenFiles.get(fd);
+    }
+    if (oe == null) {
+      LOG.error("Cannot find fd for {} in table", path);
+      return -ErrorCodes.EBADFD();
+    }
+    if (oe.getOut() != null) {
+      try {
+        oe.getOut().flush();
+      } catch (IOException e) {
+        return -ErrorCodes.EIO();
+      }
+    } else {
+      LOG.debug("Not flushing: {} was not open for writing", path);
+    }
+    return 0;
+  }
+
+  /**
+   * Retrieves file attributes.
+   * @param path The path on the FS of the file
+   * @param stat FUSE data structure to fill with file attrs
+   * @return 0 on success, negative value on error
+   */
+  @Override
+  public int getattr(String path, FileStat stat) {
+    final TachyonURI turi = mPathResolverCache.getUnchecked(path);
+    LOG.trace("getattr({}) [Tachyon: {}]", path, turi);
+    try {
+      final TachyonFile tf = mTFS.openIfExists(turi);
+      if (tf == null) {
+        return -ErrorCodes.ENOENT();
+      }
+      final FileInfo fi = mTFS.getInfo(tf);
+      stat.st_size.set(fi.getLength());
+
+      final long ctime = fi.getLastModificationTimeMs();
+      final long ctime_sec = fi.getLastModificationTimeMs() / 1000;
+      //keeps only the "residual" nanoseconds not caputred in
+      // citme_sec
+      final long ctime_nsec = (fi.getLastModificationTimeMs() % 1000) * 1000;
+      stat.st_ctim.tv_sec.set(ctime_sec);
+      stat.st_ctim.tv_nsec.set(ctime_nsec);
+      stat.st_mtim.tv_sec.set(ctime_sec);
+      stat.st_mtim.tv_nsec.set(ctime_nsec);
+
+      // Since Tachyon does not support ownership yet,
+      // set fake ownership on the user and group that
+      // is running tachyon-fuse
+      stat.st_uid.set(UID_AND_GID[0]);
+      stat.st_gid.set(UID_AND_GID[1]);
+
+      final int mode;
+      if (fi.isFolder) {
+        mode = FileStat.S_IFDIR;
+      } else {
+        mode = FileStat.S_IFREG;
+      }
+      stat.st_mode.set(mode);
+
+    } catch (InvalidPathException e) {
+      LOG.debug("Invalid path {}", path, e);
+      return -ErrorCodes.ENOENT();
+    } catch (FileDoesNotExistException e) {
+      LOG.debug("File does not exist {}", path, e);
+      return -ErrorCodes.ENOENT();
+    } catch (IOException e) {
+      LOG.error("IOException on {}", path, e);
+      return -ErrorCodes.EFAULT();
```

Comment:
It matter as long as some C program using the file system uses errno to act differetnly depending on the error. I am not sure whether this is often the case. Anyway, I fixed all the return values coming from IOExceptions to EIO for consistency.